README.md
# motorcortex-comets

## Demo

[Check it out here](https://donkeyclip.github.io/motorcortex-comets/demo/index.html)

## Installation

```bash
$ npm install --save @donkeyclip/motorcortex-comets
# OR
$ yarn add @donkeyclip/motorcortex-comets
```

## Loading

```javascript
const MotorCortex = require("@donkeyclip/motorcortex/");
const cometsDefinition = require("@donkeyclip/motorcortex-comets");
const Plugin = MotorCortex.loadPlugin(BannersDefinition);
```

# Create incident

## Comets

```javascript
const comet = new MyPlugin.Comets(
  {
    width: 800,
    height: 300,
    cometMaxSize: 581,
    cometMinSize: 200,
    meteoriteColors: ["#FFFFFF", "#FFDC63"],
    auraIner: ["#57DDBA", "#62D8B9", "#7CC8B3"],
    auraOuter: ["#47F7C5", "#87B3C1"],
    items: 3,
    repeats: 5,
    duration: 8000,
  },
  {
    selector: ".comet1",
  }
);
```

### Comets Attrs

| Name            |                 Are                  |          Values |
| --------------- | :----------------------------------: | --------------: |
| width           |          width of incident           |             num |
| height          |          height of incident          |             num |
| cometMaxSize    |      the maximum size of comets      |             num |
| cometMinSize    |      the minimum size of comets      |             num |
| meteoriteColors |   The list of colors of meteorite    | array of colors |
| auraIner        | The list of colors for the iner aura | array of colors |
| auraOuter       | The list of colors for the iner aura | array of colors |
| items           |      how meny commet will fall       |             num |
| repeats         |    how meny times clip wiil loop     |             num |
| duration        |    max duration of one comet fall    |             num |

## Stars

```javascript
const stars = new MyPlugin.Stars(
  {
    width: 800,
    height: 300,
    starMaxSize: 60,
    starMinSize: 15,
    starColor: "#F0F0F0",
    starGlowColor:
      "0 0 0 0 0.950946003   0 0 0 0 0.81265567   0 0 0 0 0.51528336  0 0 0 1 0",
    items: 15,
    repeats: 5,
    duration: 8000,
  },
  {
    selector: ".comet2",
  }
);
```

### Stars Attrs

| Name          |                     Are                     |                     Values |
| ------------- | :-----------------------------------------: | -------------------------: |
| width         |              width of incident              |                        num |
| height        |             height of incident              |                        num |
| starMaxSize   |       the maximum size of each start        |                        num |
| starMinSize   |       the minimum size of each start        |                        num |
| starColor     |             the color of a star             |                      color |
| starGlowColor |           the color of star clow            | sting of ColorMatrix value |
| items         |                start number                 |                        num |
| repeats       |        how meny times clip wiil loop        |                        num |
| duration      | max duration of one full blinking animation |                        num |

## License

[MIT License](https://opensource.org/licenses/MIT)

[<img src="https://presskit.donkeyclip.com/logos/donkey%20clip%20logo.svg" width=250></img>](https://donkeyclip.com)


CHANGELOG.md
# Changelog

## [1.1.0](https://www.github.com/donkeyclip/motorcortex-comets/compare/v1.0.6...v1.1.0) (2021-11-24)


### Features

* use CSSEffect and remove anime ([0044557](https://www.github.com/donkeyclip/motorcortex-comets/commit/0044557fb97861862e3fd630c65b72806805afee))


### Bug Fixes

* drop code from demo config ([f3110f5](https://www.github.com/donkeyclip/motorcortex-comets/commit/f3110f5ab1632d918cd549e29f3ef4fedec2660e))
* fix broken config ([277879d](https://www.github.com/donkeyclip/motorcortex-comets/commit/277879d25ae1415627a101ffb650b09d0dff4384))
* fix broken rollup config ([e68e641](https://www.github.com/donkeyclip/motorcortex-comets/commit/e68e641639005b9a6d7ddb595ce9b7cbde47e9ca))
* fix broken webpack config ([a26ddf4](https://www.github.com/donkeyclip/motorcortex-comets/commit/a26ddf4bc89f9c3e1c9f7277b41b93e270240ffd))
* update build and demo files drop old babel config ([cdfd685](https://www.github.com/donkeyclip/motorcortex-comets/commit/cdfd6859e7ea703bed298d510bd9fbcd14b630b6))
* update build and demo files drop old babel config ([9eec868](https://www.github.com/donkeyclip/motorcortex-comets/commit/9eec868e1d18ed9e9b345098be6bd83641823a9d))
* update build and demo files drop old babel config ([74390fa](https://www.github.com/donkeyclip/motorcortex-comets/commit/74390fa564af1367129a767b324c9465d4b03ee6))
* update build and demo files drop old babel config ([81e3e1c](https://www.github.com/donkeyclip/motorcortex-comets/commit/81e3e1c1609cb52f2ba9b771f67d5ef9efa24240))
* update build and demo files drop old babel config ([895fc24](https://www.github.com/donkeyclip/motorcortex-comets/commit/895fc247d50e60430d87e20545a882e794a8ebf7))
* update rollup deps to latest packages ([9d73208](https://www.github.com/donkeyclip/motorcortex-comets/commit/9d732080ee9726532db4711ab09cacc34646fde4))

### [1.0.6](https://www.github.com/donkeyclip/motorcortex-comets/compare/v1.0.5...v1.0.6) (2021-09-29)


### Bug Fixes

* update deps ([4df9627](https://www.github.com/donkeyclip/motorcortex-comets/commit/4df9627f21e95e020126c372023c1c2bf59948fe))

### [1.0.5](https://www.github.com/donkeyclip/motorcortex-comets/compare/v1.0.4...v1.0.5) (2021-09-29)


### Bug Fixes

* update npm name ([062c6b4](https://www.github.com/donkeyclip/motorcortex-comets/commit/062c6b4e863161d1e66b5c696d349371a67e0605))

### 1.0.4 (2021-05-12)


### Miscellaneous Chores

* release 1.0.4 ([20d737e](https://www.github.com/kissmybutton/motorcortex-comets/commit/20d737e3f1f006e86916ee3b827d1290b6adb6cb))


rollup.config.js
import babel from "@rollup/plugin-babel";
import commonjs from "@rollup/plugin-commonjs";
import json from "@rollup/plugin-json";
import resolve from "@rollup/plugin-node-resolve";
import { terser } from "rollup-plugin-terser";
import pkg from "./package.json";

export default [
  {
    input: "src/main.js",
    external: ["@donkeyclip/motorcortex"],
    output: [
      { file: pkg.main, format: "cjs" },
      { file: pkg.module, format: "es" },
    ],
    plugins: [resolve(), babel(), commonjs(), terser(), json()],
  },
  {
    input: "src/main.js",
    external: ["@donkeyclip/motorcortex"],
    output: [
      {
        globals: {
          "@donkeyclip/motorcortex": "MotorCortex",
        },
        name: pkg.name,
        file: pkg.browser,
        format: "umd",
      },
    ],
    plugins: [
      resolve({ mainFields: ["module", "main", "browser"] }),
      json(),
      babel(),
      commonjs(),
      terser(),
    ],
  },
];


src/Incidents/Stars.js
import { HTMLClip, CSSEffect, Combo } from "@donkeyclip/motorcortex";
import { spatial, timely } from "../helpers/randomizer";

export default class Stars extends HTMLClip {
  get html() {
    const attrs = spatial({
      dimensions: [
        [0, this.attrs.width],
        [0, this.attrs.height],
        [this.attrs.starMinSize, this.attrs.starMaxSize],
      ],
      divisions: [
        Math.floor(this.attrs.items / 2),
        Math.floor(this.attrs.items / 2),
        this.attrs.items,
      ],
      numberOfElements: this.attrs.items,
    });

    let list = [];
    for (let i = 0; i < this.attrs.items; i++) {
      const size = attrs[i][2];
      const top = attrs[i][1];
      const left = attrs[i][0];
      const comet = ` 
              <svg class="stars-svg stars-svg-${i}" width="${size}px" height="${size}px" style="left: ${left}px; top:${top}px;" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 50 50.15">
                <defs>
                  <radialGradient id="radial-gradient" cx="25.98" cy="26.84" r="22.27" gradientUnits="userSpaceOnUse">
                    <stop offset="0" stop-color="#fff9a6" stop-opacity="0.9"/>
                    <stop offset="0.07" stop-color="#fffab3" stop-opacity="0.77"/>
                    <stop offset="0.2" stop-color="#fffbc7" stop-opacity="0.57"/>
                    <stop offset="0.34" stop-color="#fffcd8" stop-opacity="0.4"/>
                    <stop offset="0.47" stop-color="#fffde6" stop-opacity="0.25"/>
                    <stop offset="0.6" stop-color="#fffef1" stop-opacity="0.14"/>
                    <stop offset="0.74" stop-color="#fffff9" stop-opacity="0.06"/>
                    <stop offset="0.87" stop-color="#fffffd" stop-opacity="0.02"/>
                    <stop offset="1" stop-color="#fff" stop-opacity="0"/>
                  </radialGradient>
                </defs>
                <title>Asset 10</title>
                <g id="Layer_2" data-name="Layer 2">
                  <g id="Layer_2-2" data-name="Layer 2">
                    <g>
                      <circle cx="25.98" cy="26.84" r="22.27" fill="url(#radial-gradient)"/>
                      <path id="stars-svg-path-1" d="M.15,26.27q13.13-2,18.73-4.52Q24.31,19.32,26,.16h0A.18.18,0,0,1,26.14,0a.19.19,0,0,1,.17.17q.6,17.88,6.81,21.58t16.72,4.62h0a.18.18,0,0,1,.16.2.19.19,0,0,1-.16.16Q38.1,27.86,33.12,32.05T26.49,50h0a.18.18,0,0,1-.36,0Q25.31,36.65,18.88,32T.17,26.62h0A.17.17,0,0,1,0,26.43.18.18,0,0,1,.15,26.27Z" fill="${this.attrs.starColor}" fill-rule="evenodd"/>
                    </g>
                  </g>
                </g>
              </svg>`;
      list += comet;
    }
    return `
    <div class="wrapper">
      ${list}
    </div>
    
    `;
  }

  get css() {
    return `
    .wrapper{
      width: ${this.attrs.width}px;
      height:${this.attrs.height}px;
      
    }
    .stars-svg{
      position: absolute;
      z-index: 3;

    }

  `;
  }

  buildTree() {
    function _getAnimatedAttr(value, opacity = false) {
      if (opacity === true) {
        return {
          opacity: value,
        };
      } else {
        return {
          transform: {
            scale: value,
          },
        };
      }
    }

    const timeData = timely({
      duration: this.attrs.duration / this.attrs.repeats,
      divisions: 3,
      numberOfElements: this.attrs.items,
      minDuration: 0.2,
      maxDuration: 1,
    });

    for (let i = 0; i < this.attrs.items; i++) {
      const starScale = new Combo(
        {
          incidents: [
            {
              incidentClass: CSSEffect,
              attrs: {
                animatedAttrs: _getAnimatedAttr(
                  0,
                  this.attrs.blinkType === "opacity"
                ),
              },
              props: {
                duration: Math.round(timeData[i][1] / 2),
              },
              position: 0,
            },
            {
              incidentClass: CSSEffect,
              attrs: {
                animatedAttrs: _getAnimatedAttr(
                  1,
                  this.attrs.blinkType === "opacity"
                ),
              },
              props: {
                duration: Math.round(timeData[i][1] / 2),
              },
              position: Math.round(timeData[i][1] / 2),
            },
          ],
        },
        {
          selector: `.stars-svg-${i}`,
          delay: timeData[i][0],
          repeats: this.attrs.repeats,
        }
      );
      this.addIncident(starScale, 0);
    }
  }
}


src/Incidents/Comets.js
import { HTMLClip, CSSEffect } from "@donkeyclip/motorcortex";
import { _addDimension, timely } from "../helpers/randomizer";
const closeAngle = 29;

export default class Comets extends HTMLClip {
  get html() {
    // first get the time (position / duration) random numbers
    this.comets = timely({
      duration: this.attrs.duration / this.attrs.repeats,
      numberOfElements: this.attrs.items,
      minDuration: 0.1,
      maxDuration: 0.9,
    });
    // console.log(JSON.parse(JSON.stringify(this.comets)));
    // secondly depending on the duration set size
    for (let i = 0; i < this.comets.length; i++) {
      const comet = this.comets[i];
      const fraction = comet[0] / (this.attrs.duration / this.attrs.repeats);
      const size =
        (this.attrs.cometMaxSize - this.attrs.cometMinSize) * (1 - fraction) +
        this.attrs.cometMinSize;
      comet.push(size); // now each row has the form [duration, position, size]
    }
    // finally we add the position random dimension. Comets can start from anyplace that will allow them to be
    // visible in our Clip even a little and even in only part of their path
    const angle = ((90 - closeAngle) * Math.PI) / 180;
    const extraWidth = this.attrs.height * Math.tan(angle);
    this.comets = _addDimension(
      {
        from: 0,
        to: this.attrs.width + extraWidth,
        numberOfElements: this.attrs.items,
      },
      this.comets
    ); // now each row has the form [duration, position, size, left]
    // console.log(this.comets);

    let list = [];
    this.itemData = [];
    const A = ((90 - closeAngle) * Math.PI) / 180; // the big angle
    const B = (closeAngle * Math.PI) / 180; // the close angle
    for (let i = 0; i < this.attrs.items; i++) {
      const size = this.comets[i][2];
      const a = size * Math.sin(B); // the horizontal size of the comet box
      const b = size * Math.sin(A); // the vertical size of the comet box
      const left = this.comets[i][3];
      this.itemData.push({ left, top: -b, width: a, size });
      const comet = ` 
                <svg class="comet-svg comet-svg-${i}" style="left: ${left}px; top: -${b}px; width: ${a}px; height: ${b}px;" xmlns="http://www.w3.org/2000/svg" class="comet-green-svg" data-name="Layer 1" viewBox="0 0 450 270.44">
                <defs>
                    <linearGradient id="b" x1="-56.99" x2="-56.74" y1="394.68" y2="394.93" gradientTransform="matrix(1363.47 0 0 -819.42 77776.05 323638.12)" gradientUnits="userSpaceOnUse">
                    <stop offset="0" stop-color="${this.attrs.auraOuter[0]}" stop-opacity=".59"/>
                    <stop offset="1" stop-color="${this.attrs.auraOuter[1]}" stop-opacity="0"/>
                    </linearGradient>
                    <linearGradient id="d" x1="-56.97" x2="-56.79" y1="394.95" y2="395.11" gradientTransform="matrix(900.62 0 0 -542.55 51377.26 214502.16)" gradientUnits="userSpaceOnUse">
                    <stop offset="0" stop-color="${this.attrs.auraIner[0]}"/>
                    <stop offset=".31" stop-color="${this.attrs.auraIner[1]}" stop-opacity=".97"/>
                    <stop offset="1" stop-color="${this.attrs.auraIner[2]}" stop-opacity="0"/>
                    </linearGradient>
                    <linearGradient id="f" x1="-56.81" x2="-56.49" y1="399.49" y2="399.76" gradientTransform="matrix(116.79 0 0 -75.59 6649.18 30453.64)" gradientUnits="userSpaceOnUse">
                    <stop offset="0" stop-color="${this.attrs.meteoriteColors[0]}"/>
                    <stop offset="1" stop-color="${this.attrs.meteoriteColors[1]}"/>
                    </linearGradient>
                    <mask id="a" width="450" height="270.44" x="0" y="0" maskUnits="userSpaceOnUse">
                    <path fill="#fff" fill-rule="evenodd" d="M11.33 230a21.52 21.52 0 0020.31 38L450 38 429.7 0 11.33 230z"/>
                    </mask>
                    <mask id="c" width="297.24" height="179.06" x="7.51" y="84.83" maskUnits="userSpaceOnUse">
                    <path fill="#fff" fill-rule="evenodd" d="M14.65 238.41a13.55 13.55 0 0012.78 23.89l277.32-153.58L292 84.83 14.65 238.41z"/>
                    </mask>
                    <mask id="e" width="38.55" height="24.95" x="12.81" y="234.29" maskUnits="userSpaceOnUse">
                    <path fill="#fff" fill-rule="evenodd" d="M17.69 241.85a9.24 9.24 0 008.73 16.29c4.5-2.41 24.94-23.85 24.94-23.85s-29.17 5.15-33.67 7.56"/>
                    </mask>
                </defs>
                <g mask="url(#a)">
                    <path fill="url(#b)" fill-rule="evenodd" d="M11.33 230a21.52 21.52 0 0020.31 38L450 38 429.7 0 11.33 230z"/>
                </g>
                <g mask="url(#c)">
                    <path fill="url(#d)" fill-rule="evenodd" d="M14.65 238.41a13.55 13.55 0 0012.78 23.89l277.32-153.58L292 84.83 14.65 238.41z"/>
                </g>
                <g mask="url(#e)">
                    <path fill="url(#f)" fill-rule="evenodd" d="M17.69 241.85a9.24 9.24 0 008.73 16.29c4.5-2.41 24.94-23.85 24.94-23.85s-29.17 5.15-33.67 7.56"/>
                </g>
                </svg>
            `;

      list += comet;
    }

    return `
            <div class="wrapper">
            ${list}
            </div>
        `;
  }

  get css() {
    return `
            .wrapper{
                width: ${this.attrs.width}px;
                height:${this.attrs.height}px;
            }
            .comet-svg{
                position: absolute;
                z-index: 3;
            }
        `;
  }

  buildTree() {
    for (let i = 0; i < this.comets.length; i++) {
      // console.log(Math.tan(angle)*(-this.itemData[i].width -this.itemData[i].left)+ this.itemData[i].top)
      const angle = (-closeAngle * Math.PI) / 180;
      const moveMagentComet = new CSSEffect(
        {
          animatedAttrs: {
            left: `-${this.itemData[i].width}px`,
            top: `${
              Math.tan(angle) *
                (-this.itemData[i].width - this.itemData[i].left) +
              this.itemData[i].top
            }px`,
          },
        },
        {
          selector: `.comet-svg-${i}`,
          duration: this.comets[i][0],
          delay: this.comets[i][1],
          repeats: this.attrs.repeats,
        }
      );
      this.addIncident(moveMagentComet, 0);
    }
  }
}


demo/index.js
import { HTMLClip, loadPlugin } from "@donkeyclip/motorcortex";
import Player from "@donkeyclip/motorcortex-player";
import MyPluginDefinition from "../dist/motorcortex-comets.esm";
const MyPlugin = loadPlugin(MyPluginDefinition);

const clip = new HTMLClip({
  html: `
    <div class="container">
      <div class="comet1"> </div>
      <div class="comet2"> </div>
    </div>`,
  css: `
    .container{
      width: 800px;
      height: 300px;
    }
  `,
  host: document.getElementById("clip"),
  containerParams: {
    width: "800px",
    height: "300px",
  },
});

const comet = new MyPlugin.Comets(
  {
    width: 800,
    height: 300,
    cometMaxSize: 1500,
    cometMinSize: 100,
    meteoriteColors: ["#FFFFFF", "#FFDC63"],
    auraIner: ["#57DDBA", "#62D8B9", "#7CC8B3"],
    auraOuter: ["#47F7C5", "#87B3C1"],
    items: 6,
    repeats: 1,
    duration: 8000,
  },
  {
    selector: ".comet1",
  }
);

const cometMagenta = new MyPlugin.Comets(
  {
    width: 800,
    height: 300,
    cometMaxSize: 1500,
    cometMinSize: 100,
    meteoriteColors: ["#FFFFFF", "#FFDC63"],
    auraIner: ["#AF4095", "#D48E93", "#CC9C9F"],
    auraOuter: ["#E06D95", "#E7B4A7"],
    items: 6,
    repeats: 1,
    duration: 8000,
  },
  {
    selector: ".comet2",
  }
);

const stars = new MyPlugin.Stars(
  {
    width: 800,
    height: 300,
    starMaxSize: 30,
    starMinSize: 5,
    starColor: "#F0F0F0",
    starGlowColor:
      "0 0 0 0 0.950946003   0 0 0 0 0.81265567   0 0 0 0 0.51528336  0 0 0 1 0",
    items: 20,
    repeats: 2,
    duration: 8000,
    blinkType: "opacity",
  },
  {
    selector: ".comet2",
  }
);

clip.addIncident(comet, 0);
clip.addIncident(cometMagenta, 0);
clip.addIncident(stars, 0);

new Player({ clip });
